subscription-fee/src/daily_operations.rs

```diff
@@ -0,0 +1,233 @@
+            );
+        }
+
+        let mut users_mapper = self.subscribed_users(service_id);
+        let energy_threshold = self.energy_threshold().get();
+        let service_info = self.service_info(service_id).get().get(service_index);
+        let current_epoch = self.blockchain().get_block_epoch();
+
+        let run_result = self.run_while_it_has_gas(GAS_TO_SAVE_PROGRESS, || {
+            if progress.current_index == total_users + 1 {
+                return STOP_OP;
+            }
+
+            let user_id = users_mapper.get_by_index(progress.current_index);
+            let opt_user_address = self.user_ids().get_address(user_id);
+            let user_address = match opt_user_address {
+                Some(address) => address,
+                None => {
+                    users_mapper.swap_remove(&user_id);
+                    total_users -= 1;
+                    return CONTINUE_OP;
+                }
+            };
+
+            progress.current_index += 1;
+
+            let subscription_type = self
+                .subscription_type(user_id, service_id, service_index)
+                .get();
+            let next_action_epoch =
+                self.get_next_action_epoch(user_id, service_id, service_index, subscription_type);
+            if next_action_epoch > current_epoch {
+                return CONTINUE_OP;
+            }
+
+            let user_energy = self.get_energy_amount(&user_address);
+            let is_premium_user = user_energy >= energy_threshold;
+            let payment_amount = if is_premium_user {
+                service_info.payment_type.amount_for_premium.clone()
+            } else {
+                service_info.payment_type.amount_for_normal.clone()
+            };
+
+            let subtract_result = self.subtract_payment(
+                user_id,
+                &service_info.payment_type.opt_specific_token,
+                &payment_amount,
+            );
+            if subtract_result.is_err() {
+                return CONTINUE_OP;
+            }
+
+            let action_results = self.perform_action(user_address.clone(), user_id, &service_info);
+            if action_results.is_err() {
+                return CONTINUE_OP;
+            }
+
+            // return funds if it didn't work? - discuss
+
+            let interpreted_results = unsafe { action_results.unwrap_unchecked() };
+            if let Some(new_token) = interpreted_results.opt_new_token {
+                self.save_new_token(user_id, new_token);
+            }
+
+            self.send_user_rewards(&user_address, interpreted_results.user_rewards);
+
+            self.last_action_epoch(user_id, service_id, service_index)
+                .set(current_epoch);
+
+            CONTINUE_OP
+        });
+
+        if run_result == OperationCompletionStatus::InterruptedBeforeOutOfGas {
+            self.save_progress(&progress);
+        }
+
+        run_result
+    }
+
+    fn subtract_payment(
```

Comment:
takeServiceFee@user@version@boolOfTransferBack - this endpoint has to be implemented at the subscription contract level. 

I have added 2 comments on safePrice integration, I think we can make the subscription contract smart enough to do the automatic swaps and to know what routes to take as staed in those comments.